### DISCUSSION TITLE
BEAM implementation in Golang

### DISCUSSION THREAD

**Discussion by halturin** [ergo-services/ergo#28] [Ideas]
There is the idea to implement BEAM virtual machine in order to run Erlang' code within Ergo node. Here is the list of advantages:
- Run Erlang' code along with Ergo' node providing transparent interaction between the Ergo and VM processes including supervision tree
- provide embeddable BEAM VM for the projects written in Golang
- use `embed` package of Golang [1.16](https://blog.golang.org/go1.16) for embedding *.erl/*.beam files. deployment Erlang code is going to be as easy as copy a single file to the destination host.
- enable Golang' profiler for the Erlang code, which I found incredibly useful. 

any thoughts? pro/cons?

**Comment by tooolbox** [comment]
As exciting as this would be, I'm curious what scale of project this is--isn't the BEAM quite complex?

**Reply by halturin** [reply]
Yeah, it's a pretty huge task, and if this proposal reaches enough demand, I will start the development. It might be I'm gonna start it as a background activity near future.

**Reply by jiait** [reply]
It's a great project and I'm looking forward to it
